File: frontend/src/lib/hdc-api/performance.ts
/**
 * 性能监控相关 API
 */

import * as App from '../../../bindings/Hadice/backend/appservice'
import type {
  CpuDetailInfo,
  MemoryDetailInfo,
  NetworkTrafficInfo,
  GraphicsInfo,
  CpuFreqInfo,
  ProcessMemoryDetail,
  FaultLogEntry,
  ProcessIOInfo,
  IpcStatInfo,
  ProcessCpuEntry
} from '@/types/hdc'

/**
 * 性能监控相关 API
 */
export const performanceAPI = {
  getCpuDetailInfo: async (connectKey: string): Promise<CpuDetailInfo> => {
    try {
      const info = await App.GetCpuDetailInfo(connectKey)
      return {
        total: info.total,
        user: info.user,
        kernel: info.kernel,
        idle: info.idle,
        iowait: info.iowait,
        irq: info.irq,
        softirq: 0, // 暂时设为 0，后端未实现
        loadAverage: {
          one: info.loadAverage?.one || 0,
          five: info.loadAverage?.five || 0,
          fifteen: info.loadAverage?.fifteen || 0
        },
        cores: [] // 暂时设为空数组，后端未实现
      }
    } catch (error) {
      console.error('[HDC API] getCpuDetailInfo failed:', error)
      throw error
    }
  },

  getProcessCpuUsage: async (connectKey: string, pid: number): Promise<ProcessCpuEntry[]> => {
    try {
      const entries = await App.GetProcessCpuUsage(connectKey, pid)
      return (entries || []).map((e) => ({
        pid: e.pid,
        totalUsage: e.totalUsage,
        userSpace: e.userSpace,
        kernelSpace: e.kernelSpace,
        name: e.name
      }))
    } catch (error) {
      console.error('[HDC API] getProcessCpuUsage failed:', error)
      throw error
    }
  },

  getMemoryDetailInfo: async (connectKey: string): Promise<MemoryDetailInfo> => {
    try {
      const info = await App.GetMemoryDetailInfo(connectKey)
      return {
        total: info.total,
        used: info.used,
        free: info.free,
        available: info.available,
        cached: info.cached,
        buffers: info.buffers,
        swapTotal: info.swapTotal,
        swapUsed: info.swapUsed,
        swapFree: info.swapFree,
        usedPercent: info.usedPercent
      }
    } catch (error) {
      console.error('[HDC API] getMemoryDetailInfo failed:', error)
      throw error
    }
  },

  getNetworkTrafficInfo: async (connectKey: string): Promise<NetworkTrafficInfo[]> => {
    try {
      const networks = await App.GetNetworkTrafficInfo(connectKey)
      return networks.map(n => ({
        interface: n.interface,
        rxBytes: n.rxBytes,
        txBytes: n.txBytes,
        rxPackets: n.rxPackets,
        txPackets: n.txPackets
      }))
    } catch (error) {
      console.error('[HDC API] getNetworkTrafficInfo failed:', error)
      throw error
    }
  },

  getGraphicsInfo: async (connectKey: string): Promise<GraphicsInfo> => {
    try {
      const info = await App.GetGraphicsInfo(connectKey)
      return {
        gpuVendor: info.gpuVendor,
        gpuRenderer: info.gpuRenderer,
        gpuVersion: info.gpuVersion,
        surfaceMemory: info.surfaceMemory,
        fpsCount: {
          fps60: info.fpsCount?.fps60 || 0,
          fps90: info.fpsCount?.fps90 || 0,
          fps120: info.fpsCount?.fps120 || 0
        },
        currentFps: 0
      }
    } catch (error) {
      console.error('[HDC API] getGraphicsInfo failed:', error)
      throw error
    }
  },

  getGraphicsInfoWithPid: async (connectKey: string, pid: number): Promise<GraphicsInfo> => {
    try {
      const info = await App.GetGraphicsInfoWithPid(connectKey, pid)
      return {
        gpuVendor: info.gpuVendor,
        gpuRenderer: info.gpuRenderer,
        gpuVersion: info.gpuVersion,
        surfaceMemory: info.surfaceMemory,
        fpsCount: {
          fps60: info.fpsCount?.fps60 || 0,
          fps90: info.fpsCount?.fps90 || 0,
          fps120: info.fpsCount?.fps120 || 0
        },
        currentFps: 0
      }
    } catch (error) {
      console.error('[HDC API] getGraphicsInfoWithPid failed:', error)
      throw error
    }
  },

  getUptimeInfo: async (connectKey: string): Promise<{ uptime: string; uptimeDays: number; loadAverage: string }> => {
    try {
      const info = await App.GetUptimeInfo(connectKey)
      return {
        uptime: info.uptime as string || 'Unknown',
        uptimeDays: info.uptimeDays as number || 0,
        loadAverage: info.loadAverage as string || '0, 0, 0'
      }
    } catch (error) {
      console.error('[HDC API] getUptimeInfo failed:', error)
      throw error
    }
  },

  getCpuFreqInfo: async (connectKey: string): Promise<CpuFreqInfo> => {
    try {
      const info = await App.GetCpuFreqInfo(connectKey)
      return {
        cores: (info.cores || []).map((c: { core: number; currentFreq: number; maxFreq: number }) => ({
          core: c.core,
          currentFreq: c.currentFreq,
          maxFreq: c.maxFreq
        }))
      }
    } catch (error) {
      console.error('[HDC API] getCpuFreqInfo failed:', error)
      throw error
    }
  },

  getProcessMemoryDetail: async (connectKey: string, pid: number): Promise<ProcessMemoryDetail> => {
    try {
      const info = await App.GetProcessMemoryDetail(connectKey, pid)
      return {
        categories: (info.categories || []).map((c: { name: string; pssTotal: number }) => ({
          name: c.name,
          pssTotal: c.pssTotal
        })),
        totalPss: info.totalPss || 0,
        swapUsed: info.swapUsed || 0,
        heapSize: info.heapSize || 0,
        heapAlloc: info.heapAlloc || 0,
        dma: info.dma || 0,
        ashmem: info.ashmem || 0
      }
    } catch (error) {
      console.error('[HDC API] getProcessMemoryDetail failed:', error)
      throw error
    }
  },

  getFaultLogList: async (connectKey: string, processName: string, n: number): Promise<FaultLogEntry[]> => {
    try {
      const entries = await App.GetFaultLogList(connectKey, processName, n)
      return (entries || []).map((e: { time: string; foreground: boolean; reason: string; recordId: string; processName: string }) => ({
        time: e.time,
        foreground: e.foreground,
        reason: e.reason,
        recordId: e.recordId,
        processName: e.processName
      }))
    } catch (error) {
      console.error('[HDC API] getFaultLogList failed:', error)
      throw error
    }
  },

  getFaultLogDetail: async (connectKey: string, recordId: string): Promise<string> => {
    try {
      return await App.GetFaultLogDetail(connectKey, recordId) || ''
    } catch (error) {
      console.error('[HDC API] getFaultLogDetail failed:', error)
      throw error
    }
  },

  getProcessIOInfo: async (connectKey: string, pid: number): Promise<ProcessIOInfo> => {
    try {
      const info = await App.GetProcessIOInfo(connectKey, pid)
      return {
        rchar: info.rchar || 0,
        wchar: info.wchar || 0,
        syscr: info.syscr || 0,
        syscw: info.syscw || 0,
        readBytes: info.readBytes || 0,
        writeBytes: info.writeBytes || 0,
        cancelledWriteBytes: info.cancelledWriteBytes || 0
      }
    } catch (error) {
      console.error('[HDC API] getProcessIOInfo failed:', error)
      throw error
    }
  },

  startIpcStat: async (connectKey: string, pid: number): Promise<void> => {
    try {
      await App.StartIpcStat(connectKey, pid)
    } catch (error) {
      console.error('[HDC API] startIpcStat failed:', error)
      throw error
    }
  },

  getIpcStat: async (connectKey: string, pid: number): Promise<IpcStatInfo> => {
    try {
      const info = await App.GetIpcStat(connectKey, pid)
      return {
        totalCount: info.totalCount || 0,
        totalTimeCost: info.totalTimeCost || 0,
        interfaces: (info.interfaces || []).map((i: { callingPid: number; descriptorCode: string; count: number; maxTime: number; minTime: number; avgTime: number }) => ({
          callingPid: i.callingPid,
          descriptorCode: i.descriptorCode,
          count: i.count,
          maxTime: i.maxTime,
          minTime: i.minTime,
          avgTime: i.avgTime
        }))
      }
    } catch (error) {
      console.error('[HDC API] getIpcStat failed:', error)
      throw error
    }
  },

  stopIpcStat: async (connectKey: string, pid: number): Promise<void> => {
    try {
      await App.StopIpcStat(connectKey, pid)
    } catch (error) {
      console.error('[HDC API] stopIpcStat failed:', error)
      throw error
    }
  },

  saveTextFile: async (filename: string, content: string): Promise<string> => {
    try {
      return await App.SaveTextFile(filename, content)
    } catch (error) {
      console.error('[HDC API] saveTextFile failed:', error)
      throw error
    }
  }
}

File: backend/performance.go
package backend

import (
	"Hadice/backend/adb"
	"Hadice/backend/device"
	"Hadice/backend/hdc"
)

// getDevicePlatform 获取设备平台类型
func getDevicePlatform(connectKey string) device.Platform {
	if deviceManager == nil {
		return device.PlatformHarmonyOS
	}
	return deviceManager.GetDevicePlatform(connectKey)
}

// GetCpuDetailInfo 获取详细 CPU 使用信息
func (a *App) GetCpuDetailInfo(connectKey string) (*hdc.CpuDetailInfo, error) {
	if getDevicePlatform(connectKey) == device.PlatformAndroid {
		return adb.GetCpuDetailInfo(connectKey)
	}
	return hdc.GetCpuDetailInfo(connectKey)
}

// GetProcessCpuUsage 获取进程级 CPU 使用率
func (a *App) GetProcessCpuUsage(connectKey string, pid int) ([]hdc.ProcessCpuEntry, error) {
	if getDevicePlatform(connectKey) == device.PlatformAndroid {
		return adb.GetProcessCpuUsage(connectKey, pid)
	}
	return hdc.GetProcessCpuUsage(connectKey, pid)
}

// GetMemoryDetailInfo 获取详细内存信息
func (a *App) GetMemoryDetailInfo(connectKey string) (*hdc.MemoryDetailInfo, error) {
	if getDevicePlatform(connectKey) == device.PlatformAndroid {
		return adb.GetMemoryDetailInfo(connectKey)
	}
	return hdc.GetMemoryDetailInfo(connectKey)
}

// GetNetworkTrafficInfo 获取网络流量信息
func (a *App) GetNetworkTrafficInfo(connectKey string) ([]hdc.NetworkTrafficInfo, error) {
	if getDevicePlatform(connectKey) == device.PlatformAndroid {
		return adb.GetNetworkTrafficInfo(connectKey)
	}
	return hdc.GetNetworkTrafficInfo(connectKey)
}

// GetGraphicsInfo 获取图形性能信息
func (a *App) GetGraphicsInfo(connectKey string) (*hdc.GraphicsInfo, error) {
	if getDevicePlatform(connectKey) == device.PlatformAndroid {
		return adb.GetGraphicsInfo(connectKey)
	}
	return hdc.GetGraphicsInfo(connectKey)
}

// GetGraphicsInfoWithPid 获取图形性能信息（带 PID，用于安卓 FPS）
func (a *App) GetGraphicsInfoWithPid(connectKey string, pid int) (*hdc.GraphicsInfo, error) {
	if getDevicePlatform(connectKey) == device.PlatformAndroid {
		return adb.GetGraphicsInfoWithPid(connectKey, pid)
	}
	return hdc.GetGraphicsInfo(connectKey)
}

// GetUptimeInfo 获取系统运行时间信息
func (a *App) GetUptimeInfo(connectKey string) (map[string]interface{}, error) {
	if getDevicePlatform(connectKey) == device.PlatformAndroid {
		return adb.GetUptimeInfo(connectKey)
	}
	return hdc.GetUptimeInfo(connectKey)
}

// GetCpuFreqInfo 获取 CPU 频率信息
func (a *App) GetCpuFreqInfo(connectKey string) (*hdc.CpuFreqInfo, error) {
	if getDevicePlatform(connectKey) == device.PlatformAndroid {
		return adb.GetCpuFreqInfo(connectKey)
	}
	return hdc.GetCpuFreqInfo(connectKey)
}

// GetProcessMemoryDetail 获取进程内存详情
func (a *App) GetProcessMemoryDetail(connectKey string, pid int) (*hdc.ProcessMemoryDetail, error) {
	if getDevicePlatform(connectKey) == device.PlatformAndroid {
		return adb.GetProcessMemoryDetail(connectKey, pid)
	}
	return hdc.GetProcessMemoryDetail(connectKey, pid)
}

// GetFaultLogList 获取故障日志列表
func (a *App) GetFaultLogList(connectKey string, processName string, n int) ([]hdc.FaultLogEntry, error) {
	if getDevicePlatform(connectKey) == device.PlatformAndroid {
		return adb.GetFaultLogList(connectKey, processName, n)
	}
	return hdc.GetFaultLogList(connectKey, processName, n)
}

// GetFaultLogDetail 获取故障日志详情
func (a *App) GetFaultLogDetail(connectKey string, recordId string) (string, error) {
	if getDevicePlatform(connectKey) == device.PlatformAndroid {
		return adb.GetFaultLogDetail(connectKey, recordId)
	}
	return hdc.GetFaultLogDetail(connectKey, recordId)
}

// GetProcessIOInfo 获取进程 IO 信息
func (a *App) GetProcessIOInfo(connectKey string, pid int) (*hdc.ProcessIOInfo, error) {
	if getDevicePlatform(connectKey) == device.PlatformAndroid {
		return adb.GetProcessIOInfo(connectKey, pid)
	}
	return hdc.GetProcessIOInfo(connectKey, pid)
}

// StartIpcStat 开始 IPC 统计
func (a *App) StartIpcStat(connectKey string, pid int) error {
	if getDevicePlatform(connectKey) == device.PlatformAndroid {
		return nil
	}
	return hdc.StartIpcStat(connectKey, pid)
}

// GetIpcStat 获取 IPC 统计数据
func (a *App) GetIpcStat(connectKey string, pid int) (*hdc.IpcStatInfo, error) {
	if getDevicePlatform(connectKey) == device.PlatformAndroid {
		return &hdc.IpcStatInfo{}, nil
	}
	return hdc.GetIpcStat(connectKey, pid)
}

// StopIpcStat 停止 IPC 统计
func (a *App) StopIpcStat(connectKey string, pid int) error {
	if getDevicePlatform(connectKey) == device.PlatformAndroid {
		return nil
	}
	return hdc.StopIpcStat(connectKey, pid)
}
